token-metadata/program/src/processor/verification/collection_verification.rs

```diff
@@ -0,0 +1,218 @@
+use crate::{
+    assertions::{
+        assert_owned_by, collection::assert_collection_verify_is_valid,
+        metadata::assert_metadata_derivation,
+    },
+    error::MetadataError,
+    instruction::{Context, MetadataDelegateRole, Unverify, Verify},
+    state::{AuthorityRequest, AuthorityType, Metadata, TokenMetadataAccount},
+    utils::{clean_write_metadata, decrement_collection_size, increment_collection_size},
+};
+use mpl_utils::assert_signer;
+use solana_program::{entrypoint::ProgramResult, pubkey::Pubkey};
+
+pub(crate) fn verify_collection_v1(program_id: &Pubkey, ctx: Context<Verify>) -> ProgramResult {
+    // Assert program ownership/signers.
+
+    // Authority account is the collection authority and must be a signer.
+    assert_signer(ctx.accounts.authority_info)?;
+
+    if let Some(delegate_record_info) = ctx.accounts.delegate_record_info {
+        assert_owned_by(delegate_record_info, program_id)?;
+    }
+
+    assert_owned_by(ctx.accounts.metadata_info, program_id)?;
+
+    let collection_mint_info = ctx
+        .accounts
+        .collection_mint_info
+        .ok_or(MetadataError::MissingCollectionMint)?;
+    assert_owned_by(collection_mint_info, &spl_token::ID)?;
+
+    let collection_metadata_info = ctx
+        .accounts
+        .collection_metadata_info
+        .ok_or(MetadataError::MissingCollectionMetadata)?;
+    assert_owned_by(collection_metadata_info, program_id)?;
+
+    let collection_master_edition_info = ctx
+        .accounts
+        .collection_master_edition_info
+        .ok_or(MetadataError::MissingCollectionMasterEdition)?;
+    assert_owned_by(collection_master_edition_info, program_id)?;
+
+    // Deserialize item metadata and collection parent metadata.
+    let mut metadata = Metadata::from_account_info(ctx.accounts.metadata_info)?;
+    let mut collection_metadata = Metadata::from_account_info(collection_metadata_info)?;
+
+    // Don't verify already verified items, otherwise for sized collections we end up with invalid
+    // size data.
+    if let Some(collection) = &metadata.collection {
+        if collection.verified {
+            return Err(MetadataError::AlreadyVerified.into());
+        }
+    }
+
+    // Verify the collection in the item's metadata matches the collection mint.  Also verify
+    // the collection metadata matches the collection mint, and the collection edition derivation.
+    assert_collection_verify_is_valid(
```

Comment:
I think there's a larger refactor you can do here to only do the match on the metadata.collection once. [I made a PR to your branch for it](https://github.com/metaplex-foundation/metaplex-program-library/pull/1042/files) to make it easier to see the diff. If you like it, you can just merge it into your branch, otherwise you can close it after you've taken a look.

Reply:
I missed that `assert_master_edition` is doing the supply check. Yeah, we want to avoid weirdness by ensuring that Collection Parents are not print editions and are master editions that cannot print so I think you're right that this account needs to stay.